### ISSUE TITLE
Cache-line aligned allocation for OpenMP

### ISSUE THREAD

**Issue by Slaedr** [ginkgo-project/ginkgo#920]
Currently, the OpenMP `raw_alloc` just uses `malloc`. Typically we would want to have allocated memory blocks aligned to cache-line boundaries, typically 64 bytes. Two things would be needed:
- [ ] Query the cache line length during configuration. I believe most modern CPUs use 64 bytes, but this would be good to have.
- [ ] A simple (non-interface) class to implement the alignment logic (or simply use [std::align](https://en.cppreference.com/w/cpp/memory/align)) and delete it properly in the end. Adding to alignment to `OmpExecutor::raw_alloc` instead is not a nice option because it's protected, and `Executor::alloc` has logging that is not thread-safe.

Especially, the semantics we need is allocation of a group of memory blocks, the beginning of each of which is aligned to 64-byte boundaries. Typically, this can be done by aligning the beginning of the large block and making the stride a multiple of 64.

The current use case on my mind is dynamic "shared memory" for batched openmp solvers.

**Comment by upsj** [comment]
Our system is not able to do that portably (and may not until C++ supports it natively) since you can't reconstruct the original pointer returned from `malloc` after it has been aligned.

**Comment by Slaedr** [comment]
Yes, that's why I propose to keep the original pointer in a struct/class and delete using it once we are done. For now, I'm thinking of a static scope-based object that can return the aligned pointer when needed, and it would have a destructor that would use the cached original pointer to delete the entire memory block.

**Comment by Slaedr** [comment]
Wait, actually, `aligned_alloc` [is there](https://en.cppreference.com/w/c/memory/aligned_alloc) in the C11 standard, and C++ 17 just "inherits" from that. Maybe we can just use the C11 version while still compiling with C++14. I think this works for the alignment logic and properly freeing the memory, and could be used in `OmpExecutor::raw_alloc`. `aligned_alloc` has a slightly greater overhead than simple `malloc`, though I'm positive the benefits outweigh that. In addition, I would add a small function for allocating a large block of memory containing many sub-blocks each aligned to a cache line.

**Comment by upsj** [comment]
With C++17, we could extend #1315 to also enable aligned allocation.

**Comment by upsj** [comment]
We can query the L1 cache size using the "appropriately" named [`hardware_destructive_interference_size`](https://en.cppreference.com/w/cpp/thread/hardware_destructive_interference_size)
